index.js

```diff
@@ -1,310 +1,333 @@
-const express = require('express')
-const https = require('https')
-const fs = require('fs');
-const cors = require('cors')
-const path = require('path');
-const app = express()
-const port = 8000
+const express = require("express");
+const https = require("https");
+const fs = require("fs");
+const cors = require("cors");
+const path = require("path");
+const app = express();
+const port = 8000;
+const crypto = require("crypto");
+const dotenv = require("dotenv");
 
-app.use(cors());
-app.use(express.json());
+// Enable environment variables
+dotenv.config({
+  path: "./.env",
+});
 
-// Front end endpoints
-app.get('/', (req, res) => {
-  res.send('Hello World!')
-})
+// Parse JSON bodies
+app.use(express.json());
 
-app.get('/auth', (req, res) => {
-  console.log("Auth Happened!")
-  res.sendFile(path.join(__dirname, '/auth.html'))
-})
+// Enable CORS (https://developers.asana.com/docs/security)
+app.use(cors());
 
-// This function runs before every api request. It validates the request.
+// (run before every API request)
 app.use((req, res, next) => {
-  // Check Expires
-  const expires = req.query.expires || req.body.expires
-  if (new Date().getTime() > expires) {
-    console.log("Request Expired.")
-    return
+  // Assess timeliness (https://developers.asana.com/docs/timeliness)
+  const expirationDate = req.query.expires_at || req.body.expires_at;
+  
+  if (new Date().getTime() > expirationDate) {
+    console.log("Request expired.");
+    return;
   }
 
-  // For the example app, we can't share the signature's secret with you. For now, let's just validate the signature exists
-
-  // Check Signature
+  // Assess message integrity (https://developers.asana.com/docs/message-integrity)
   if (!req.headers["x-asana-request-signature"]) {
-    console.log("Signature Missing.")
-    return
+    console.log("Signature is missing.");
+    return;
+  }
+
+  let stringToVerify;
+
+  if (req.method === "POST") {
+    stringToVerify = req.body.data.toString();
+  } else if (req.method === "GET") {
+    stringToVerify = req._parsedUrl.query;
+  }
+
+  let computedSignature = crypto
+    .createHmac("sha256", process.env.CLIENT_SECRET)
```

Comment:
Might be helpful to include a comment indicating where to get this client secret (from the dev console)

Reply:
Actually, since the configuration of this app is owned by us, a developer pulling this code would have no way of setting this env variable. This means that the app would not be in a working state until this signature verification is removed. Probably makes sense to leave this commented out, and a short explanation of why (CLIENT_SECRET isn't accessible). Still would be nice to describe how to use this code with an app owned by a developer.

Reply:
Good call. The Client Secret is tied to the app, which we own. I'll comment this out and add notes.